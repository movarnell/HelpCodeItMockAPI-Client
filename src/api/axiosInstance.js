// src/api/axiosInstance.js
import axios from 'axios';

const axiosInstance = axios.create({
  baseURL: 'http://localhost:5100', // Adjust if necessary
});

axiosInstance.interceptors.request.use(
    (config) => {
      const token = localStorage.getItem('token');
      console.log('Token from localStorage:', token);
      if (token) {
        config.headers['Authorization'] = `Bearer ${token}`;
      } else {
        console.warn('No token found in localStorage');
      }
      return config;
    },
    (error) => {
      return Promise.reject(error);
    }
  );

export default axiosInstance;