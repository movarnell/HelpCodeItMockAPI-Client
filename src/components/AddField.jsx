import React, { useState } from 'react';
import axiosInstance from '../api/axiosInstance';
import { useNavigate, useParams, Link } from 'react-router-dom';
import Navigation from './Navigation';
import { IoArrowBack } from 'react-icons/io5'; // Import the back arrow icon

function AddField() {
  const { id } = useParams(); // Endpoint ID
  const [fieldName, setFieldName] = useState('');
  const [dataType, setDataType] = useState('VARCHAR');
  const [isRequired, setIsRequired] = useState(false);
  const [error, setError] = useState('');
  const navigate = useNavigate();

  const dataTypes = [
    'INT',
    'VARCHAR',
    'TEXT',
    'DATE',
    'DATETIME',
    'BOOLEAN',
    'FLOAT',
  ];

  const handleSubmit = async (e) => {
    e.preventDefault();
    try {
      await axiosInstance.post(`/endpoints/${id}/fields`, {
        fields: [
          {
            field_name: fieldName.trim(),
            data_type: dataType,
            is_required: isRequired,
          },
        ],
      });
      navigate(`/endpoints/${id}/fields`);
    } catch (err) {
      setError(err.response.data.message || 'Failed to add field');
    }
  };

  const handleNavigate = (path) => {
    navigate(path);
  };

  return (
    <div className="min-h-screen bg-gray-100">
      <Navigation />
      <div className="py-6 mx-auto max-w-7xl sm:px-6 lg:px-8">
        <div className="flex items-center mb-6">
          <Link
            to="/endpoints"
            className="mr-4 text-indigo-600 transition-colors duration-200 hover:text-indigo-800"
          >
            <IoArrowBack className="w-8 h-8" />
          </Link>
          <h1 className="text-3xl font-bold text-gray-900">Add Field</h1>
        </div>

        <p className="mb-4 text-sm text-gray-600">
          Note: You don't need to assign an ID - the API will automatically generate and assign an ID for each new field.
        </p>

        {error && <p className="mb-4 text-red-600">{error}</p>}
        <div className="overflow-hidden bg-white shadow sm:rounded-lg">
          <form onSubmit={handleSubmit} className="p-6 space-y-6">
            <div>
              <label htmlFor="fieldName" className="block text-sm font-medium text-gray-700">Field Name</label>
              <input
                type="text"
                id="fieldName"
                className="block w-full px-3 py-2 mt-1 border border-gray-300 rounded-md shadow-sm focus:outline-none focus:ring-indigo-500 focus:border-indigo-500 sm:text-sm"
                value={fieldName}
                onChange={(e) => setFieldName(e.target.value.trim())}
                required
              />
            </div>
            <div>
              <label htmlFor="dataType" className="block text-sm font-medium text-gray-700">Data Type</label>
              <select
                id="dataType"
                className="block w-full py-2 pl-3 pr-10 mt-1 text-base border-gray-300 rounded-md focus:outline-none focus:ring-indigo-500 focus:border-indigo-500 sm:text-sm"
                value={dataType}
                onChange={(e) => setDataType(e.target.value)}
              >
                {dataTypes.map((type) => (
                  <option key={type} value={type}>{type}</option>
                ))}
              </select>
            </div>
            <div className="flex items-start">
              <div className="flex items-center h-5">
                <input
                  id="isRequired"
                  type="checkbox"
                  className="w-4 h-4 text-indigo-600 border-gray-300 rounded focus:ring-indigo-500"
                  checked={isRequired}
                  onChange={(e) => setIsRequired(e.target.checked)}
                />
              </div>
              <div className="ml-3 text-sm">
                <label htmlFor="isRequired" className="font-medium text-gray-700">Is Required</label>
              </div>
            </div>
            <button type="submit" className="flex justify-center w-full px-4 py-2 text-sm font-medium text-white bg-indigo-600 border border-transparent rounded-md shadow-sm hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500">
              Add Field
            </button>
          </form>
        </div>
        <div className="mt-8 space-x-4">
          <button
            onClick={() => handleNavigate(`/endpoints/${id}/add-data`)}
            className="inline-flex items-center px-4 py-2 text-sm font-medium text-white bg-green-600 border border-transparent rounded-md hover:bg-green-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-green-500"
          >
            Add Data
          </button>
          <Link to={'/endpoints'} className="inline-flex items-center px-4 py-2 text-sm font-medium text-white bg-blue-600 border border-transparent rounded-md hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-blue-500">Back to Endpoints</Link>
          <Link to="/dashboard" className="inline-flex items-center px-4 py-2 text-sm font-medium text-white bg-blue-600 border border-transparent rounded-md hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-blue-500">
            Back to Dashboard
          </Link>
        </div>
      </div>
    </div>
  );
}

export default AddField;