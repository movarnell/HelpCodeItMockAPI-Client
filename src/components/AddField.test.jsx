import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { BrowserRouter, useParams } from 'react-router-dom';
import AddField from './AddField';
import axiosInstance from '../api/axiosInstance';
import { act } from 'react-dom/test-utils';

jest.mock('../api/axiosInstance');
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => jest.fn(),
  useParams: jest.fn(),
}));

describe('AddField Component', () => {
  beforeEach(() => {
    useParams.mockReturnValue({ id: '1' });
  });

  const renderComponent = async () => {
    await act(async () => {
      render(
        <BrowserRouter>
          <AddField />
        </BrowserRouter>
      );
    });
  };

  test('renders add field form', async () => {
    await renderComponent();
    expect(screen.getByLabelText('Field Name')).toBeInTheDocument();
    expect(screen.getByLabelText('Data Type')).toBeInTheDocument();
    expect(screen.getByLabelText('Is Required')).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /add field/i })).toBeInTheDocument();
  });

  test('submits form with field data', async () => {
    axiosInstance.post.mockResolvedValue({ data: { field_id: 1 } });
    await renderComponent();

    await act(async () => {
      fireEvent.change(screen.getByLabelText('Field Name'), { target: { value: 'test_field' } });
      fireEvent.change(screen.getByLabelText('Data Type'), { target: { value: 'STRING' } });
      fireEvent.click(screen.getByLabelText('Is Required'));
      fireEvent.click(screen.getByRole('button', { name: /add field/i }));
    });

    await waitFor(() => {
      expect(axiosInstance.post).toHaveBeenCalledWith('/endpoints/1/fields', {
        fields: [{
          field_name: 'test_field',
          data_type: 'STRING',
          is_required: true,
        }]
      });
    });
  });
});