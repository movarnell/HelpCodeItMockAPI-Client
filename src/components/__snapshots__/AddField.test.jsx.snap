// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddField Component matches snapshot 1`] = `
<div
  className="min-h-screen bg-gray-100"
>
  <nav
    className="bg-white shadow"
  >
    <div
      className="px-4 mx-auto max-w-7xl sm:px-6 lg:px-8"
    >
      <div
        className="flex justify-between h-16"
      >
        <div
          className="flex"
        >
          <div
            className="flex items-center flex-shrink-0"
          >
            <a
              className="text-2xl font-bold text-indigo-600"
              href="/dashboard"
              onClick={[Function]}
            >
              Help Mock It 
            </a>
          </div>
          <div
            className="hidden sm:ml-6 sm:flex sm:space-x-8"
          >
            <a
              className="inline-flex items-center px-1 pt-1 text-sm font-medium border-b-2 border-transparent text-gray-500 hover:border-gray-300 hover:text-gray-700"
              href="/dashboard"
              onClick={[Function]}
            >
              Dashboard
            </a>
            <a
              className="inline-flex items-center px-1 pt-1 text-sm font-medium border-b-2 border-transparent text-gray-500 hover:border-gray-300 hover:text-gray-700"
              href="/endpoints"
              onClick={[Function]}
            >
              Endpoints
            </a>
            <a
              className="inline-flex items-center px-1 pt-1 text-sm font-medium border-b-2 border-transparent text-gray-500 hover:border-gray-300 hover:text-gray-700"
              href="/endpoints/create"
              onClick={[Function]}
            >
              Create Endpoint
            </a>
          </div>
        </div>
        <div
          className="flex items-center"
        >
          <button
            className="inline-flex items-center px-4 py-2 ml-3 text-sm font-medium text-white bg-indigo-600 border border-transparent rounded-md hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500"
            onClick={[Function]}
          >
            Logout
          </button>
        </div>
      </div>
    </div>
  </nav>
  <div
    className="py-6 mx-auto max-w-7xl sm:px-6 lg:px-8"
  >
    <div
      className="flex items-center mb-6"
    >
      <a
        className="mr-4 text-indigo-600 transition-colors duration-200 hover:text-indigo-800"
        href="/endpoints"
        onClick={[Function]}
      >
        <svg
          className="w-8 h-8"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M244 400 100 256l144-144M120 256h292"
            fill="none"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="48"
          />
        </svg>
      </a>
      <h1
        className="text-3xl font-bold text-gray-900"
      >
        Add Field
      </h1>
    </div>
    <p
      className="mb-4 text-sm text-gray-600"
    >
      Note: You don't need to assign an ID - the API will automatically generate and assign an ID for each new field.
    </p>
    <div
      className="overflow-hidden bg-white shadow sm:rounded-lg"
    >
      <form
        className="p-6 space-y-6"
        onSubmit={[Function]}
      >
        <div>
          <label
            className="block text-sm font-medium text-gray-700"
            htmlFor="fieldName"
          >
            Field Name
          </label>
          <input
            className="block w-full px-3 py-2 mt-1 border border-gray-300 rounded-md shadow-sm focus:outline-none focus:ring-indigo-500 focus:border-indigo-500 sm:text-sm"
            id="fieldName"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
        </div>
        <div>
          <label
            className="block text-sm font-medium text-gray-700"
            htmlFor="dataType"
          >
            Data Type
          </label>
          <select
            className="block w-full py-2 pl-3 pr-10 mt-1 text-base border-gray-300 rounded-md focus:outline-none focus:ring-indigo-500 focus:border-indigo-500 sm:text-sm"
            id="dataType"
            onChange={[Function]}
            value="VARCHAR"
          >
            <option
              value="INT"
            >
              INT
            </option>
            <option
              value="VARCHAR"
            >
              VARCHAR
            </option>
            <option
              value="TEXT"
            >
              TEXT
            </option>
            <option
              value="DATE"
            >
              DATE
            </option>
            <option
              value="DATETIME"
            >
              DATETIME
            </option>
            <option
              value="BOOLEAN"
            >
              BOOLEAN
            </option>
            <option
              value="FLOAT"
            >
              FLOAT
            </option>
          </select>
        </div>
        <div
          className="flex items-start"
        >
          <div
            className="flex items-center h-5"
          >
            <input
              checked={false}
              className="w-4 h-4 text-indigo-600 border-gray-300 rounded focus:ring-indigo-500"
              id="isRequired"
              onChange={[Function]}
              type="checkbox"
            />
          </div>
          <div
            className="ml-3 text-sm"
          >
            <label
              className="font-medium text-gray-700"
              htmlFor="isRequired"
            >
              Is Required
            </label>
          </div>
        </div>
        <button
          className="flex justify-center w-full px-4 py-2 text-sm font-medium text-white bg-indigo-600 border border-transparent rounded-md shadow-sm hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500"
          type="submit"
        >
          Add Field
        </button>
      </form>
    </div>
    <div
      className="mt-8 space-x-4"
    >
      <button
        className="inline-flex items-center px-4 py-2 text-sm font-medium text-white bg-green-600 border border-transparent rounded-md hover:bg-green-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-green-500"
        onClick={[Function]}
      >
        Add Data
      </button>
      <a
        className="inline-flex items-center px-4 py-2 text-sm font-medium text-white bg-blue-600 border border-transparent rounded-md hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-blue-500"
        href="/endpoints"
        onClick={[Function]}
      >
        Back to Endpoints
      </a>
      <a
        className="inline-flex items-center px-4 py-2 text-sm font-medium text-white bg-blue-600 border border-transparent rounded-md hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-blue-500"
        href="/dashboard"
        onClick={[Function]}
      >
        Back to Dashboard
      </a>
    </div>
    // Display existing fields with delete option
    <div
      className="mt-6"
    >
      <h2
        className="text-2xl font-semibold"
      >
        Existing Fields
      </h2>
      <ul />
    </div>
  </div>
</div>
`;
